src/dimension/broadcast.rs

```diff
@@ -0,0 +1,115 @@
+use crate::error::*;
+use crate::{Dimension, Ix0, Ix1, Ix2, Ix3, Ix4, Ix5, Ix6, IxDyn};
+
+/// Calculate the co_broadcast shape of two dimensions. Return error if shapes are
+/// not compatible.
+fn broadcast_shape<D1, D2, Output>(shape1: &D1, shape2: &D2) -> Result<Output, ShapeError>
+where
+    D1: Dimension,
+    D2: Dimension,
+    Output: Dimension,
+{
+    let (k, overflow) = shape1.ndim().overflowing_sub(shape2.ndim());
+    // Swap the order if d2 is longer.
+    if overflow {
+        return broadcast_shape::<D2, D1, Output>(shape2, shape1);
+    }
+    // The output should be the same length as shape1.
+    let mut out = Output::zeros(shape1.ndim());
+    let out_slice = out.slice_mut();
+    let s1 = shape1.slice();
+    let s2 = shape2.slice();
+    // Uses the [NumPy broadcasting rules]
+    // (https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html#general-broadcasting-rules).
+    //
+    // Zero dimension element is not in the original rules of broadcasting.
+    // We currently treat it as the same as 1. Especially, when one side is
+    // zero with one side is empty, or both sides are zero, the result will
+    // remain zero.
+    for i in 0..shape1.ndim() {
+        out_slice[i] = s1[i];
+    }
+    for i in 0..shape2.ndim() {
+        if out_slice[i + k] != s2[i] && s2[i] != 0 {
+            if out_slice[i + k] <= 1 {
+                out_slice[i + k] = s2[i]
+            } else if s2[i] != 1 {
+                return Err(from_kind(ErrorKind::IncompatibleShape));
+            }
+        }
+    }
+    Ok(out)
+}
+
+pub trait BroadcastShape<Other: Dimension> {
+    /// The resulting dimension type after broadcasting.
+    type BroadcastOutput: Dimension;
```

Comment:
A trait with an "output" type (See RawDataSubst and FnOnce as inspiration) should almost always prefer the associated type name "Output" as a matter of convention. Think of it as a type-level function's return value.

Reply:
Thanks much for teaching.  I changed BroadCastOutput to Output